src/main/java/duke/storage/Storage.java

```diff
@@ -0,0 +1,101 @@
+package duke.storage;
+
+import duke.exception.DukeException;
+import duke.task.Deadline;
+import duke.task.Event;
+import duke.task.Task;
+import duke.task.Todo;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Scanner;
+
+/**
+ * The object to read and store the file.
+ */
+public class Storage {
+    private String filePath; // The file's path.
+
+    /**
+     * Initiates the object.
+     *
+     * @param filePath The file's path.
+     */
+    public Storage(String filePath) {
+        this.filePath = filePath;
+    }
+
+    /**
+     * Loads the file.
+     *
+     * @return Tasks loaded from the file.
+     * @throws DukeException When cannot found file or format incorrect.
+     * @throws FileNotFoundException Would not happen.
+     */
+    public ArrayList<Task> load() throws DukeException, FileNotFoundException {
+        ArrayList<Task> tasks = new ArrayList<Task>();
+
+        File f = new File(filePath);
+        if (!f.exists()) {
+            throw new DukeException("File not found.");
+        }
+
+        Scanner readFileScanner = new Scanner(f);
+        while (readFileScanner.hasNext()) {
+            String[] todoTask = readFileScanner.nextLine().split(" \\| ");
+            switch (todoTask[0]) {
+            case "T":
+                Task task1 = new Todo(todoTask[2]);
+                if (todoTask[1].equals("1")) {
+                    task1.markAsDone();
+                }
+                tasks.add(task1);
+                break;
+            case "D":
+                Task task2 = new Deadline(todoTask[2], todoTask[3]);
+                if (todoTask[1].equals("1")) {
+                    task2.markAsDone();
+                }
+                tasks.add(task2);
+                break;
+            case "E":
+                Task task3 = new Event(todoTask[2], todoTask[3]);
+                if (todoTask[1].equals("1")) {
+                    task3.markAsDone();
+                }
+                tasks.add(task3);
+                break;
+            default:
+                throw new DukeException("Something in file go wrong.");
+            }
+        }
+        readFileScanner.close();
+
+        return tasks;
+    }
+
+    /**
+     * Saves tasks to file.
+     *
+     * @param tasks The tasks to save.
+     * @throws DukeException When file does not exist.
+     */
+    public void saveFile(ArrayList<Task> tasks) throws DukeException {
```

Comment:
A few things I noticed here:

1. Method Argument could use the more generic `List<Task>` rather than a specific class `ArrayList<Task>`. This modification helps the method to accept more subclasses of `List<T>` as arguments.

2. If program terminates (assume an unhandled exception) without ExitCommand.java::execute() being called, my changes would not be saved. An easy fix would be to add the following to your Duke.java:run() try-catch block:

```
try {
...
finally {
    ui.showLine();
    database.update(taskList);
}
```
3. Should my list size grow to the thousands (though unlikely), the list would take longer to write to file.

For 3. you could write an append(int index, String data) method that appends the new task into the file. Since each line denotes one task, and they are in running order, it would be easy for us to "insert" a new task. This reduces our write operation to a single line rather than n lines.

Then calling the new append(int index, String data) after each task is added to the list, would then eliminate the issues in 2 and 3! There is a tradeoff however, there will be more IO operations.